### PR TITLE
added hooks support

### PR DESCRIPTION
**Pull request by rehat101**
added hooks support

added new APIs:
`Ctx.Provider`
`Ctx.use()`

### CODE REVIEW

File: src/CtxHooks.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  getState() {
+    return this.state;
+  }
+}
+
+function createUseSelector(Context) {
+  return function useSelector(selector) {
+    const store = React.useContext(Context);
+    const [, forceUpdate] = React.useState(0);
```
Comment (brigand) [inline src/CtxHooks.js]: Should be `useState(false)` if it's going to end up being a boolean when line 44 runs.

**Comment by rehat101** [comment]
Also need to add prop validations and error handlers to make the api more bullet proof

File: src/CtxHooks.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    let selectedState;
+
+    React.useEffect(() => {
+      const unsub = store.subscribe(selector);
+      forceUpdate(x => !x);
```
Comment (brigand) [inline src/CtxHooks.js]: this should be:

```js
const unsub = store.subscribe((state) => {
  if (selector produces a different value from last time) {
    forceUpdate(x => !x);
  }
});
```

(point being that the forceUpdate happens inside the subscribe callback)

File: src/CtxHooks.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return selectedState;
+  };
+}
+
+function makeCtx() {
+  const store = new CtxStore();
+  const Context = React.createContext();
+
+  return {
+    provider: function CtxComponent({ set, children }) {
```
Comment (brigand) [inline src/CtxHooks.js]: Name should be 'Provider' as it's a component. We can probably also rename 'set' to 'value' here.

File: src/CtxHooks.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    const store = React.useContext(Context);
+    const [, forceUpdate] = React.useState(0);
+    let selectedState;
+
+    React.useEffect(() => {
+      const unsub = store.subscribe(selector);
+      forceUpdate(x => !x);
+      // detach when unmount
+      return () => unsub();
+    }, [store]);
```
Comment (brigand) [inline src/CtxHooks.js]: This also depends on 'selector', and I don't think 'store' can ever change so it's not technically needed in the dependency array.

File: src/CtxHooks.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    this.state = newState;
+  }
+
+  notifyAll() {
+    this.subs.forEach(sub => sub());
+  }
+}
+
+function createUseSelector(Context) {
+  const execSelector = (value, selector) => {
```
Comment (brigand) [inline src/CtxHooks.js]: This could be placed outside of `createUseSelector`

File: src/CtxHooks.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    if (typeof selector === 'function') {
+      return selector(value);
+    } else if (selector === null || selector === undefined) {
+      return value;
+    } else {
+      throw new Error('Invalid selector');
```
Comment (brigand) [inline src/CtxHooks.js]: Could be more clear with "Expected the selector to be a function or null/undefined"

File: src/CtxHooks.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        const newState = execSelector(state, selector);
+
+        if (newState !== prevState.current) {
+          prevState.current = newState;
+          forceUpdate(x => !x);
+        }
+      });
+
+      // detach when unmount
+      return () => unsub();
```
Comment (brigand) [inline src/CtxHooks.js]: This could just be `return unsub`

File: src/CtxHooks.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    return prevState.current;
+  };
+}
+
+function makeCtx() {
+  const store = new CtxStore();
+  const Context = React.createContext();
+
+  return {
+    Provider: function CtxComponent({ value, children }) {
```
Comment (brigand) [inline src/CtxHooks.js]: Probably name it `CtxProvider` for better stack traces.

File: src/CtxHooks.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    }, [selector]);
+
+    prevState.current = execSelector(store.state, selector);
+
+    return prevState.current;
+  };
+}
+
+function makeCtx() {
+  const store = new CtxStore();
```
Comment (brigand) [inline src/CtxHooks.js]: I didn't notice until now, but we should actually have the store created inside `Provider`

```js
const store = React.useMemo(() => new CtxStore(), [])
```

File: src/__tests__/Ctx-hooks-test.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import { render } from '@testing-library/react';
+import { makeCtx } from '../CtxHooks';
+
+
+it(`works`, () => {
+  const Ctx = makeCtx();
+
+  function B() {
+    const x = Ctx.use(state => state.x);
+    return <p>Hello {x}</p>;
```
Comment (brigand) [inline src/__tests__/Ctx-hooks-test.js]: Maybe something less casual than greeting a letter, like `<p>state.x: {x}</p>`

File: src/__tests__/Ctx-hooks-test.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  class A extends React.Component {
+    render() {
+      return (
+        <Ctx.Provider value={{ x: this.props.x, y: 'a' }}>
```
Comment (brigand) [inline src/__tests__/Ctx-hooks-test.js]: 'y' is never used. Could be removed.

**Pull request merged**